src/lib/MessageCache.js

```diff
@@ -0,0 +1,285 @@
+'use strict';
+
+function bsearchMin(a,b,test){
+  // b is true and valid
+  while(b-a>1){
+    var mid = Math.floor((a+b)/2);
+    test(mid) ? b = mid : a = mid
+  }
+  return b;
+}
+
+function bsearchMax(a,b,test){
+  // a is true and valid
+  while(b-a>1){
+    var mid = Math.floor((a+b)/2);
+    test(mid) ? a = mid : b = mid
+  }
+  return a;
+}
+
+class User {
+
+  constructor(userId, details){
+    this.id = userId;
+    this.timestamp = [];
+    this.details = details;
+  }
+
+  lastTimestamp(){
+    let l = this.timestamp.length;
+    if( l === 0 ) return null;
+    return this.timestamp[l-1];
+  }
+
+  addTimestamp(t){
+    // t is unix time accurate to second
+    this.timestamp.push(t);
+  }
+
+  numMsgBetween(start, end){
+    let ts = this.timestamp;
+    let len = ts.length;
+    if( len === 0 || ts[len-1] < start || end < ts[0] ) return 0;
+
+    let mi = bsearchMin(-1,len-1,function(i){
+      return start <= ts[i];
+    });
+    let mx = bsearchMax(0,len, function(i){
+      return ts[i] <= end;
+    });
+    return mx-mi+1;
+  }
+
+  name(){
+    let d = this.details;
+    if( d.first_name && d.last_name ) return d.first_name + ' ' + d.last_name;
+    if( d.first_name ) return d.first_name;
+    if( d.last_name ) return d.last_name;
+    if( d.username ) return d.username;
+    return '';
+  }
+
+  sort(){
+    this.timestamp.sort( (a,b) => a-b );
```

Comment:
This looks like an insertion sort to me, not sure which algorithm is used in nodejs?